feat(core): Only consider non-readonly properties in request bodies

maikdijkstra opened the pull request:
## Status

<!--- **READY/WIP/HOLD** --->

**READY**

## Description

Whenever a readonly property is defined in a request body then these may be sent as part of a response but should not be sent as part of a request according to the openapi 3.0 specification. To solve this using types, we can only allow the properties that are non-readonly since these are the actual request body.

Should fix: #616

## Steps to Test or Reproduce

Outline the steps to test or reproduce the PR here.

1. Use a readonly property in a requestBody
2. Check that the readonly properties are discarded from the generated body type in the defined hooks.

anymaniax commented on packages/orval/src/client.ts:
```diff
@@ -35,6 +36,13 @@ export const GENERATOR_CLIENT: GeneratorClients = {
   zod: zod()(),
 };
 
+const ORVAL_DEPENDENCIES: GeneratorDependency[] = [
```
I don't know about importing orval as a dependency of the generated code. You create then a direct dependence on orval. Would make more sense to generate it automatically with all the other schemas no?

maikdijkstra replied:
Yes, I think that would be better specifically if Orval is only a dev dependency. The only question that I then have how should we distinguish this "orval defined" schema from "specification defined" schemas? Because I don't know whether it would be acceptable to mix these types among all other schemas?

What do you think would be a proper place to put these types @anymaniax?

anymaniax replied:
Good question because in that case, we could have some conflicts with the generated models from the specification. Do you know that you can already achieve that with a mutator and exporting a BodyType in it? 

``` ts
export type BodyType<Body> = NonReadonly<Body>;
```

maikdijkstra replied:
I did not know this then technically speaking the issue can be solved by simply setting the mutator correctly.
```typescript
// https://stackoverflow.com/questions/49579094/typescript-conditional-types-filter-out-readonly-properties-pick-only-requir/49579497#49579497
type IfEquals<X, Y, A = X, B = never> = (<T>() => T extends X ? 1 : 2) extends <
  T,
>() => T extends Y ? 1 : 2
  ? A
  : B;

type WritableKeys<T> = {
  [P in keyof T]-?: IfEquals<
    { [Q in P]: T[P] },
    { -readonly [Q in P]: T[P] },
    P
  >;
}[keyof T];

type Writable<T> = Pick<T, WritableKeys<T>>;
export type NonReadonly<T> = {
  [P in keyof Writable<T>]: T[P] extends number | string | boolean
    ? T[P]
    : NonReadonly<NonNullable<T[P]>>;
};

export type BodyType<Body> = NonReadonly<Body>;
```

maikdijkstra replied:
Is this small use-case then important enough to change Orval to adhere to this OpenAPI specification? Or should this be the solution to issue #616 ?

maikdijkstra commented:
@anymaniax I updated the PR, so now we keep track of whether the type contains any readonly properties, if it contains a readonly property anywhere in the object tree then the ```NonReadonly``` type is added to the imports at the request body. So, when readonly is not there than we also dont import it. We add the orval generated types in the index.ts of the schemas. Let me know whether this would be an appropriate solution.

anymaniax commented:
@maikdijkstra thanks a lot. Seems great. I found it weird to inject it in the index.ts. What do you think about inject it in each body file that need it? I was thinking would this need to be deactivated for some case? I don’t think so but what is your thoughts about this? I thought also about another solution. Be able to do the same thing as with a mutator but without the need to define all the mutator.

maikdijkstra commented:
@anymaniax My reason to inject it into the index.ts is that it will not show up with any other schemas in a separate file but is importable from the schemas folder and therefore only generate this type once. Does the body file mean the file with the endpoints? If so, then yes that is also an option, you will get quite some duplication if you generate an endpoint per file and all of them need it.

I think this case should not be deactivatable, since then the specification is simply wrong according to the OpenApi specification. 

I don't really follow your proposed solution using the mutator. Could you explain how that would work?

maikdijkstra commented:
@anymaniax Good to hear! I added a test case to the default config. I also updated the react-query/basic sample to have a readonly property in the request body

anymaniax commented:
Super @maikdijkstra I plan to checkout and test the branch tomorrow and merge it after that. I will create a new release with it during the week.

anymaniax commented:
@maikdijkstra for me all is good when you have time can you check the comment of @localrobot

michaelslec commented:
Will we get the same functionality for `writeonly` fields?
